### ISSUE TITLE
Questions about the difference with DDPO

### ISSUE THREAD

**Issue by DannyXYZa** [XueZeyue/DanceGRPO#72]
Hi, I have a question about the difference with DDPO. For (1) I think the objective is very similar as in DanceGRPO the kl in the original GRPO is removed. (2) I think DDPO also support computing the advantages within the prompt-level group (https://github.com/kvablack/ddpo-pytorch/blob/1958463f020112c9a7bc85768d296daacc2e1b4b/scripts/train.py#L427) ? 

Also, may I ask how do you implement the DDPO for SDE RF in the paper results?

Thank you so much!

<img width="1585" height="1064" alt="Image" src="https://github.com/user-attachments/assets/393de40f-77ed-4091-a00a-b627c114a686" />

**Comment by XueZeyue** [comment]
> > For Policy Gradient, DDPO implements a **Prompt Tracker** (https://github.com/kvablack/ddpo-pytorch/blob/main/ddpo_pytorch/stat_tracking.py#L5), and here is the basic logic:
> > 
> > 1. Tracks statistics per prompt: Maintains a running buffer of rewards for each unique prompt to compute mean and standard deviation.
> > 2. Computes advantages: Normalizes rewards using prompt-specific statistics (**if sufficient data exists**) or global batch statistics (**otherwise**) to produce advantages.
> > 
> > For each unique prompt:
> > 
> > 1. Retrieve all rewards for that prompt.
> > 2. Update the prompt's reward buffer (deque) with new rewards.
> > 3. If the buffer has fewer than **min_count (always set to 8 or 16)** rewards, use the global mean/std of the current batch's rewards.
> > 4. Otherwise, compute the prompt-specific mean/std from the buffer.
> > 
> > **This implementation presents a challenge: when the prompt set is large and diverse—for instance, around 30,000 prompts—and we use 300 iterations with a batch size of 96, it is likely that many prompts in the dataset will not be sampled even once, let alone meet the minimum count requirement. As a result, we have to rely on the global mean and standard deviation of the rewards from the current batch.** Actually, if we use the global mean and standard deviation, the training will fail in less than 10 iterations.
> > I followed the original DDPO framework to train models such as FLUX (as illustrated) and Seedream over the past year. However, after a month without successful experiments, I switched to the GRPO approach. This method involves generating multiple rollouts per prompt in each iteration, which reduces the number of distinct prompts seen per iteration but significantly improves training stability.
> > There is also a previous work discussing the instability of DDPO in large-scale prompt sets (PRDP,https://arxiv.org/abs/2402.08714), but they follow the paradigm of DPO to improve it, rather than Policy Gradient. Policy Gradient is much more efficient than DPO.
> 
> You said you I switched to the GRPO approach. I find the code is PPO, not GRPO at all. There isn't any KL module. It is totally the same loss in DDPO. I mean DDPO use PPO. You use GRPO-KL=PPO, so the loss is the same. https://github.com/XueZeyue/DanceGRPO/blob/d97950b51def6e61fddda83b0dbcbc615b07997c/fastvideo/train_grpo_sd.py#L652C25-L652C30

I think you misunderstood GRPO and PPO.
The key difference between GRPO and PPO is that GRPO omits the critic function and uses rollouts to simulate the critic, rather than KL-divergence.
PPO is a typical actor-critic RL algorithm, utilizing the critic to calculate the value function and employing the policy gradient to update the policy.
